### PR TITLE
Documentation for diagnostic codes used in repo

### LINKED ISSUE #47287: Consider in-repo documentation for diagnostic codes used in repo
Follow-up from https://github.com/dotnet/aspnetcore/pull/47118#discussion_r1131714773.

The runtime repo documents diagnostic codes used by their libraries in: https://github.com/dotnet/runtime/blob/main/docs/project/list-of-diagnostics.md

We currently use diagnostic codes in several places in the repo:

- Source generators
- MVC analyzers
- Framework analyzers

But don't have a central location that documents all of them.

### PR DESCRIPTION
**Pull request by divyeshio**
# Documentation for diagnostic codes used in repo

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

We currently use diagnostic codes in several places in the repo:

Source generators
MVC analyzers
Framework analyzers
But don't have a central location that documents all of them.

- This PR is created to document all the diagnostic codes used throughout the repo in one place. Similar to what runtime repo does.

Fixes #47287

### CODE REVIEW

File: docs/list-of-diagnostics.md
Comment (captainsafia) [inline docs/list-of-diagnostics.md:1]: Can we add this file to the table of contents in https://github.com/dotnet/aspnetcore/blob/main/docs/README.md?

**Comment by divyeshio** [comment]
- I did a directory wide search for ASP0002, but couldn't find it.
- Diagnostic codes starting with **BL** are present in two places : [here](https://github.com/dotnet/aspnetcore/blob/542da423ff3394e86fd7608fc425fadd33ba3fd9/src/Components/Analyzers/src/DiagnosticDescriptors.cs) and [here](https://github.com/dotnet/aspnetcore/blob/542da423ff3394e86fd7608fc425fadd33ba3fd9/src/Tools/SDK-Analyzers/Components/src/DiagnosticDescriptors.cs) which I think is reasoned in https://github.com/dotnet/aspnetcore/issues/10382 but BL0007 is not present in of them, not sure if this was intended or not.

**Comment by captainsafia** [comment]
> * I did a directory wide search for ASP0002, but couldn't find it.
> * Diagnostic codes starting with **BL** are present in two places : [here](https://github.com/dotnet/aspnetcore/blob/542da423ff3394e86fd7608fc425fadd33ba3fd9/src/Components/Analyzers/src/DiagnosticDescriptors.cs) and [here](https://github.com/dotnet/aspnetcore/blob/542da423ff3394e86fd7608fc425fadd33ba3fd9/src/Tools/SDK-Analyzers/Components/src/DiagnosticDescriptors.cs) which I think is reasoned in [Rationalize Razor Compiler diagnostic IDs #10382](https://github.com/dotnet/aspnetcore/issues/10382) but BL0007 is not present in one of them, not sure if this was intended or not.

Good discoveries that this process reveals! My suspicion is the gaps in the diagnostic IDs are a result of user error (AKA creating a new analyzer and skipped a value) as opposed to intentional issues.

**Review by captainsafia — APPROVED** [review]

**Review by mkArtakMSFT — APPROVED** [review]
Thanks for your help, @divyeshio !

**Comment by mkArtakMSFT** [comment]
@captainsafia is there anything you're waiting for here ? If not, feel free to merge.

**Comment by danmoseley** [comment]
Thanks @divyeshio .. looks like this is your 2nd PR here. We hope you will continue to contribute -- we have [over 100 marked "help wanted". ](https://github.com/dotnet/aspnetcore/labels/help%20wanted) - let us know if you need help finding a particular type of issue to work on.

**Comment by captainsafia** [comment]
Nope, this is good to merge! I'll in fact be making use of this in a current PR of mine. :)

cc: @eerhardt

**Pull request merged**